chainer/functions/matmul.py

```diff
@@ -0,0 +1,205 @@
+            return gx0, gx1
+
+
+def matmul(a, b, transa=False, transb=False):
+    """Compute matrix multiplication of two arrays.
+
+    Args:
+        a, b: Variables of 2-D or 1-D arrays.
+            A 2-D array with shape (N, M) is considered as a NxM matrix.
+            A 1-D array with shape (N,) is considered as a Nx1 matrix.
+        transa (bool): If true, transpose a.
+        transb (bool): If true, transpose b.
+
+    Returns:
+        ~chainer.Variable: Matrix multiplication as a 2-D array
+    """
+    return MatMul(transa=transa, transb=transb)(a, b)
+
+
+class BatchMatMul(function.Function):
+    def __init__(self, transa=False, transb=False):
+        self.transa = transa
+        self.transb = transb
+
+    def _output_shape(self, a, b):
+        batch_size = a.shape[0]
+        a_mat_shape = _as_mat(a[0]).shape
+        b_mat_shape = _as_mat(b[0]).shape
+        m = a_mat_shape[1] if self.transa else a_mat_shape[0]
+        n = b_mat_shape[0] if self.transb else b_mat_shape[1]
+        return (batch_size, m, n)
+
+    def forward_cpu(self, x):
+        a, b = x
+        assert a.shape[0] == b.shape[0]
+        batch_size = a.shape[0]
+        shape = self._output_shape(a, b)
+        ret = numpy.empty(shape)
+        for i in six.moves.range(batch_size):
+            ret[i] = _matmul_cpu(
+                a[i], b[i], transa=self.transa, transb=self.transb)
+        return ret,
+
+    def backward_cpu(self, x, gy):
+        a, b = x
+        batch_size = a.shape[0]
+        ga = numpy.empty(a.shape)
+        gb = numpy.empty(b.shape)
+        for i in six.moves.range(batch_size):
+            ga[i] = _matmul_cpu(gy[0][i], b[i],
+                                transb=not self.transb, transout=self.transa
+                                ).reshape(a[0].shape)
+            gb[i] = _matmul_cpu(a[i], gy[0][i],
+                                transa=not self.transa, transout=self.transb
+                                ).reshape(b[0].shape)
+        return ga, gb
+
+    def forward_gpu(self, x):
+        a, b = x
+        assert a.shape[0] == b.shape[0]
+        shape = self._output_shape(a, b)
+        ret = cuda.empty(shape)
+        _batch_matmul_gpu(a, b,
+                          transa=self.transa, transb=self.transb, out=ret)
+        return ret,
+
+    def backward_gpu(self, x, gy):
+        a, b = x
+        batch_size = a.shape[0]
+        ga = cuda.empty((batch_size,) + _as_mat(a[0]).shape)
+        gb = cuda.empty((batch_size,) + _as_mat(b[0]).shape)
+        _batch_matmul_gpu(gy[0], b,
+                          transb=not self.transb, transout=self.transa, out=ga)
+        _batch_matmul_gpu(a, gy[0],
+                          transa=not self.transa, transout=self.transb, out=gb)
+        ga = ga.reshape(a.shape)
+        gb = gb.reshape(b.shape)
+        return ga, gb
+
+
+def batch_matmul(a, b, transa=False, transb=False):
+    """Compute matrix multiplication of two arrays in a batch manner.
```

Comment:
-> Compute**s** **the** matrix multiplication**s** of two **sets of** arrays in a batch manner.